File: examples/python/grpc_service/server_lib.py
"""GreeterServicer implementation — importable for testing."""
import time

import greeter_pb2
import greeter_pb2_grpc


class GreeterServicer(greeter_pb2_grpc.GreeterServicer):
    def SayHello(self, request, context):
        return greeter_pb2.HelloReply(message=f"Hello, {request.name}!")

    def SayHelloStream(self, request, context):
        for i in range(5):
            yield greeter_pb2.HelloReply(message=f"Hello #{i + 1}, {request.name}!")
            time.sleep(0.1)

File: examples/go/grpc_service/server_lib.go
package grpcservice

import (
	"context"
	"fmt"
	"time"

	pb "cross_platform/examples/proto/greeter"
)

type GreeterServer struct {
	pb.UnimplementedGreeterServer
}

func (s *GreeterServer) SayHello(_ context.Context, req *pb.HelloRequest) (*pb.HelloReply, error) {
	return &pb.HelloReply{Message: fmt.Sprintf("Hello, %s!", req.Name)}, nil
}

func (s *GreeterServer) SayHelloStream(req *pb.HelloRequest, stream pb.Greeter_SayHelloStreamServer) error {
	for i := range 5 {
		if err := stream.Send(&pb.HelloReply{Message: fmt.Sprintf("Hello #%d, %s!", i+1, req.Name)}); err != nil {
			return err
		}
		time.Sleep(100 * time.Millisecond)
	}
	return nil
}
